### PR TITLE
[DNSSEC] Allow using of a custom OpenSSL engine for BIND

### PR DESCRIPTION
**Pull request by stanislavlevin**
For now Debian, Fedora, RHEL, etc. build BIND with 'native PKCS11'
support. Till recently, that was the strict requirement of DNSSEC.
The problem is that this restricts cross-platform features of FreeIPA.

With the help of libp11, which provides `pkcs11` engine plugin for
the OpenSSL library for accessing PKCS11 modules in a semi-
transparent way, FreeIPA could utilize OpenSSL version of BIND.

BIND in turn provides ability to specify the OpenSSL engine on the
command line of `named` and all the BIND `dnssec-*` tools by using
the `-E engine_name`.

Related: https://pagure.io/freeipa/issue/8094

### CODE REVIEW

**Comment by abbra** [comment]
This looks good as a base. We need to add more, though:
- a method to migrate bind-pkcs11 to bind configuration
- update of enabled services

For migrating Fedora / RHEL, we need:
- new libp11 and softhsm packages in the target distributions
- change to fedora/rhel platform code to enable openssl configuration
- new dependencies in the spec file so that we obsolete bind-pkcs11 and require standard bind

File: ipaserver/install/dnskeysyncinstance.py [hunk 1]
```diff
@@ -186,23 +187,57 @@ def __setup_softhsm(self):
...
+            os.fchmod(f.fileno(), 0o644)
+            f.write(softhsm_conf_txt)
+
+        if constants.NAMED_SSL_ENGINE is not None:
+            # setup OpenSSL config file
+            # this one is needed because FreeIPA installation
+            # disables p11-kit-proxy PKCS11 module
+            # https://github.com/OpenSC/libp11
+            openssl_conf_txt = textwrap.dedent("""
+                # OpenSSL configuration file
```
Comment (tiran) [inline ipaserver/install/dnskeysyncinstance.py]: I don't like to have hard-coded configuration files hidden in the middle of lot of code. Please create a template in ``install/share/`` and use it here.

Comment (stanislavlevin) [reply ipaserver/install/dnskeysyncinstance.py]: That was the easiest way.
I don't insist, will be done in the next patch version.

**Comment by stanislavlevin** [comment]
@tiran, thank you for the early warnings.

**Comment by abbra** [comment]
@stanislavlevin could you please rebase to latest master?

**Comment by abbra** [comment]
Current status: use of custom OpenSSL version of BIND fails in FIPS mode.

**Comment by tiran** [comment]
Named is failing to load the root key. The out-of-memory error message usually means that an algorithm is blocked by FIPS.

```
Dec 16 14:45:00 master.ipa.test named[24079]: none:107: 'max-cache-size 90%' - setting to 2406MB (out of 2674MB)
Dec 16 14:45:00 master.ipa.test named[24079]: /etc/named.root.key:12: configuring managed key for '.': out of memory
Dec 16 14:45:00 master.ipa.test named[24079]: loading configuration: failure
Dec 16 14:45:00 master.ipa.test named[24079]: exiting (due to fatal error)
```

**Comment by stanislavlevin** [comment]
Oops, I missed this year comments. Sorry.

File: freeipa.spec.in [hunk 2]
```diff
@@ -100,6 +100,16 @@
 %global httpd_version 2.4.41-6.1
 %endif
 
+# BIND in Fedora 31 and later uses 'pkcs11' OpenSSL engine instead of native
+# PKCS11
+%if 0%{?fedora} >= 31
+    %global with_bind_pkcs11 0
+    %global openssl_pkcs11_version 0.4.10-3
```
Comment (tiran) [inline freeipa.spec.in:107]: Please require ``0.4.10-6`` with fix for https://bugzilla.redhat.com/show_bug.cgi?id=1827535

**Comment by stanislavlevin** [comment]
Rebased.

File: ipaplatform/base/constants.py [hunk 3]
```diff
@@ -23,6 +23,8 @@ class BaseConstantsNamespace:
     NAMED_USER = "named"
     NAMED_GROUP = "named"
     NAMED_DATA_DIR = "data/"
+    NAMED_OPTIONS_VAR = "OPTIONS"
+    NAMED_SSL_ENGINE = None
```
Comment (tiran) [inline ipaplatform/base/constants.py:27]: The name ``NAMED_SSL_ENGINE`` feels wrong. It's not about SSL but about OpenSSL or cryptographic operations. How about ``NAMED_CRYPTO_ENGINE`` or ``NAMED_OPENSSL_ENGINE``?

File: ipaplatform/fedora/paths.py [hunk 4]
```diff
@@ -36,6 +37,9 @@ class FedoraPathNamespace(RedHatPathNamespace):
     NAMED_CRYPTO_POLICY_FILE = "/etc/crypto-policies/back-ends/bind.config"
     if HAS_NFS_CONF:
         SYSCONFIG_NFS = '/etc/nfs.conf'
+    if HAS_BIND_OPENSSL_ENGINE:
+        DNSSEC_OPENSSL_CONF = "/etc/ipa/dnssec/openssl.cnf"
+        DNSSEC_KEYFROMLABEL = "/usr/sbin/dnssec-keyfromlabel"
```
Comment (tiran) [inline ipaplatform/fedora/paths.py:42]: ``DNSSEC_KEYFROMLABEL`` is only defined in Fedora namespace and not used anywhere.

File: ipaserver/install/dnskeysyncinstance.py [hunk 5]
```diff
@@ -231,23 +277,17 @@ def __setup_softhsm(self):
...
-        named_fd = open(paths.DNSSEC_SOFTHSM_PIN_SO, 'w')
-        named_fd.seek(0)
-        named_fd.truncate(0)
-        named_fd.write(pin_so)
-        named_fd.close()
-        # owner must be root
-        os.chmod(paths.DNSSEC_SOFTHSM_PIN_SO, 0o400)
+        with open(paths.DNSSEC_SOFTHSM_PIN_SO, 'w') as f:
+            # owner must be root
+            os.chmod(f.fileno(), 0o400)
```
Comment (tiran) [inline ipaserver/install/dnskeysyncinstance.py:289]: ``os.fchmod()``

File: ipaserver/install/server/upgrade.py [hunk 6]
```diff
@@ -946,8 +946,7 @@ def named_root_key_include():
 
 
 def named_update_global_forwarder_policy():
-    bind = bindinstance.BindInstance()
-    if not bindinstance.named_conf_exists() or not bind.is_configured():
+    if not bindinstance.named_conf_exists():
```
Comment (tiran) [inline ipaserver/install/server/upgrade.py:949]: The function name ``named_conf_exists`` is confusing. It checks that ``/etc/named.conf`` exists and is configured for IPA. How about you pass in ``bind`` instance to all functions and use ``bind.is_configured()`` instead?

File: ipaserver/install/server/upgrade.py [hunk 7]
```diff
@@ -1469,30 +1491,44 @@ def uninstall_dogtag_9(ds, http):
     http.restart()
 
 
-def mask_named_regular():
-    """Disable named, we need to run only named-pkcs11, running both named and
-    named-pkcs can cause unexpected errors"""
-    if sysupgrade.get_upgrade_state('dns', 'regular_named_masked'):
+def named_switch_service():
```
Comment (tiran) [inline ipaserver/install/server/upgrade.py:1494]: A test is failing due to an error in this function:

```
2020-04-29T06:57:55Z ERROR IPA server upgrade failed: Inspect /var/log/ipaupgrade.log and run command ipa-server-upgrade manually.
2020-04-29T06:57:55Z DEBUG   File "/usr/lib/python3.7/site-packages/ipapython/admintool.py", line 179, in execute
    return_value = self.run()
  File "/usr/lib/python3.7/site-packages/ipaserver/install/ipa_server_upgrade.py", line 54, in run
    server.upgrade()
  File "/usr/lib/python3.7/site-packages/ipaserver/install/server/upgrade.py", line 2407, in upgrade
    upgrade_configuration()
  File "/usr/lib/python3.7/site-packages/ipaserver/install/server/upgrade.py", line 2206, in upgrade_configuration
    named_switch_service(),
  File "/usr/lib/python3.7/site-packages/ipaserver/install/server/upgrade.py", line 1526, in named_switch_service
    bind.setup_service()
  File "/usr/lib/python3.7/site-packages/ipaserver/install/bindinstance.py", line 800, in setup_service
    self.ldap_configure('DNS', self.fqdn, None, self.suffix)
  File "/usr/lib/python3.7/site-packages/ipaserver/install/service.py", line 654, in ldap_configure
    CONFIGURED_SERVICE, name, fqdn, ldap_suffix, config
  File "/usr/lib/python3.7/site-packages/ipaserver/install/service.py", line 671, in _ldap_enable
    post_add_config=extra_config_opts)
  File "/usr/lib/python3.7/site-packages/ipaserver/install/service.py", line 177, in set_service_entry_config
    api.Backend.ldap2.add_entry(entry)
  File "/usr/lib/python3.7/site-packages/ipapython/ipaldap.py", line 1648, in add_entry
    self.conn.add_s(str(entry.dn), list(attrs.items()))
  File "/usr/lib64/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/lib/python3.7/site-packages/ipapython/ipaldap.py", line 1086, in error_handler
    raise errors.NotFound(reason=arg_desc or 'no such entry')

2020-04-29T06:57:55Z DEBUG The ipa-server-upgrade command failed, exception: NotFound: no such entry
2020-04-29T06:57:55Z ERROR Unexpected error - see /var/log/ipaupgrade.log for details:
```

**Comment by stanislavlevin** [comment]
@tiran, thank you for the review, I'll try to check this today.

**Comment by abbra** [comment]
@stanislavlevin could you please rebase? Overall, the changes work (tested in bind-dyndb-ldap CI with bind 9.16), I think we should start merging this.

File: ipaplatform/base/constants.py (hunk 3 above)
Comment (stanislavlevin) [reply ipaplatform/base/constants.py:27]: Sure, renamed `NAMED_SSL_ENGINE` `NAMED_OPENSSL_ENGINE`.

File: ipaplatform/fedora/paths.py (hunk 4 above)
Comment (stanislavlevin) [reply ipaplatform/fedora/paths.py:42]: ```DNSSEC_KEYFROMLABEL = "/usr/sbin/dnssec-keyfromlabel-pkcs11" ``` in BasePathNamespace

File: ipaserver/install/dnskeysyncinstance.py (hunk 5 above)
Comment (stanislavlevin) [reply ipaserver/install/dnskeysyncinstance.py:289]: Fixed.

File: ipaserver/install/server/upgrade.py (hunk 6 above)
Comment (stanislavlevin) [reply ipaserver/install/server/upgrade.py:949]: `bind.is_configured()` is the link to `bindinstance.named_conf_exists()`. Perhaps, to speed up the checking of that condition the module scope function was added (no need to instantiate of `BindInstance`).
Each `named_` function in the upgrade code uses the same logic:
```
 542     if not bindinstance.named_conf_exists():                                    
 543         # DNS service may not be configured                                     
 544         logger.info('DNS is not configured')                                    
 545         return False
```
This results in each of these functions read named.conf. I'll move this check out of the scope of `named_*` and check only once.

**Comment by stanislavlevin** [comment]
I'm working on this.

**Comment by stanislavlevin** [comment]
If a vendor supplies `named` with custom command-line options then this configuration will be overridden by IPA installer. I'm afraid that I can't correctly parse that.

**Comment by stanislavlevin** [comment]
Checking if Ubuntu has such env variable.

**Comment by stanislavlevin** [comment]
Ubuntu 16.04, 18.04:
```
EnvironmentFile=/etc/default/bind9                 
OPTIONS="-u bind"
```
Ubuntu 20.04:
```
EnvironmentFile=/etc/default/named                 
OPTIONS="-u bind"
```
This means I have to add parsing of options.
@tjaalton, which sysconf should I add?

**Comment by stanislavlevin** [comment]
I guess it is ready for review.

File: ipaserver/install/server/upgrade.py (hunk 7 above)
Comment (stanislavlevin) [reply ipaserver/install/server/upgrade.py:1423]: Fixed

File: ipaplatform/fedora/constants.py [hunk 8]
```diff
@@ -16,6 +16,8 @@
 # Fedora 30 and later use /etc/nfs.conf
 # Fedora 29 has both
 HAS_NFS_CONF = osinfo.version_number >= (30,)
+# BIND employs 'pkcs11' OpenSSL engine instead of native PKCS11
+HAS_BIND_OPENSSL_ENGINE = osinfo.version_number >= (31,)
```
Comment (tiran) [inline ipaplatform/fedora/constants.py:20]: You can simplify the extra checks and assume that Fedora has OpenSSL engien support. Fedora 30 is no longer supported.

File: ipaplatform/base/paths.py [hunk 9]
```diff
@@ -243,8 +244,6 @@ class BasePathNamespace:
     IPA_REPLICA_CONNCHECK = "/usr/sbin/ipa-replica-conncheck"
     IPA_RMKEYTAB = "/usr/sbin/ipa-rmkeytab"
     IPACTL = "/usr/sbin/ipactl"
-    NAMED = "/usr/sbin/named"
-    NAMED_PKCS11 = "/usr/sbin/named-pkcs11"
```
Comment (tiran) [inline ipaplatform/base/paths.py:247]: Memo to me: We need a way to formerly deprecate constants before we remove them.

File: ipaserver/install/bindinstance.py [hunk 10]
```diff
@@ -990,11 +984,8 @@ def __setup_named_conf(self):
         named_txt = ipautil.template_file(
             os.path.join(paths.USR_SHARE_IPA_DIR, "bind.named.conf.template"),
             self.sub_dict)
-        named_fd = open(paths.NAMED_CONF, 'w')
-        named_fd.seek(0)
-        named_fd.truncate(0)
-        named_fd.write(named_txt)
-        named_fd.close()
+        with open(paths.NAMED_CONF, 'w') as named_fd:
+            named_fd.write(named_txt)
```
Comment (tiran) [inline ipaserver/install/bindinstance.py:988]:
```suggestion
        with open(paths.NAMED_CONF, 'w') as f:
            os.fchmod(f.fileno(), 0o600)
            os.fchown(f.fileno(), 0, self.named_gid)
            f.write(named_txt)
```

File: ipatests/test_ipaserver/test_install/test_bindinstance.py [hunk 11]
```diff
@@ -78,19 +78,3 @@ def test_add_crypto_policy(m_set, m_get, namedconf):
...
-    # IPA related settings.
-    m_get.return_value = False
-    with open(namedconf, 'w') as f:
-        f.write(TEST_CONFIG)
-
-    result = named_add_crypto_policy()
-    assert not result
-
-    m_get.assert_not_called()
-    m_set.assert_not_called()
```
Comment (tiran) [inline ipatests/test_ipaserver/test_install/test_bindinstance.py:96]: Why did you remove this test case?

File: ipaplatform/fedora/constants.py (hunk 8 above)
Comment (stanislavlevin) [reply ipaplatform/fedora/constants.py:20]: ok, done.

File: ipaserver/install/bindinstance.py (hunk 10 above)
Comment (stanislavlevin) [reply ipaserver/install/bindinstance.py:988]: 600?

Comment (stanislavlevin) [reply ipaserver/install/bindinstance.py:988]: Permissions are under the control of BIND package (platform-specific) and usually are `640,root,named`.
But to be sure they are correct we can explicitly set ones.

Comment (stanislavlevin) [reply ipaserver/install/bindinstance.py:988]: Fixed.

File: ipatests/test_ipaserver/test_install/test_bindinstance.py (hunk 11 above)
Comment (stanislavlevin) [reply ipatests/test_ipaserver/test_install/test_bindinstance.py:96]: I moved this change out to its own commit.

Comment (stanislavlevin) [reply ipatests/test_ipaserver/test_install/test_bindinstance.py:96]: `test_add_crypto_policy_no_ipa` checks if migration was skipped if named.conf was not IPA-configured, after refactoring this check is performed for all the `named_*` migration functions.

File: ipaplatform/base/paths.py (hunk 9 above)
Comment (stanislavlevin) [reply ipaplatform/base/paths.py:247]: Good idea!

File: ipaplatform/fedora/paths.py [hunk 12]
```diff
@@ -36,6 +36,8 @@ class FedoraPathNamespace(RedHatPathNamespace):
     NAMED_CRYPTO_POLICY_FILE = "/etc/crypto-policies/back-ends/bind.config"
     if HAS_NFS_CONF:
         SYSCONFIG_NFS = '/etc/nfs.conf'
+    DNSSEC_OPENSSL_CONF = "/etc/ipa/dnssec/openssl.cnf"
+    DNSSEC_KEYFROMLABEL = "/usr/sbin/dnssec-keyfromlabel"
```
Comment (tjaalton) [inline ipaplatform/fedora/paths.py:40]: shouldn't these be in base/paths.py?

Comment (stanislavlevin) [reply ipaplatform/fedora/paths.py:40]: They are already in
https://github.com/freeipa/freeipa/blob/b65fed7c9de754ac653543c61ab725bad3a0e1e1/ipaplatform/base/paths.py#L235
https://github.com/freeipa/freeipa/blob/b65fed7c9de754ac653543c61ab725bad3a0e1e1/ipaplatform/base/paths.py#L70

Comment (stanislavlevin) [reply ipaplatform/fedora/paths.py:40]: If you mean defaults, then I think the defaults should be RHEL based.
Maybe I'm wrong.

File: ipaserver/install/bindinstance.py (hunk 10 above)
Comment (tiran) [reply ipaserver/install/bindinstance.py]: sorry, ``0o640`` is right

File: ipatests/test_ipaserver/test_install/test_bindinstance.py (hunk 11 above)
Comment (tiran) [reply ipatests/test_ipaserver/test_install/test_bindinstance.py:96]: Thanks that makes sense!

File: ipaplatform/base/paths.py (hunk 9 above)
Comment (stanislavlevin) [reply ipaplatform/base/paths.py:247]: if we consider `ipaplatform` as the part of the public API then I'm wrong.

File: ipaplatform/fedora/paths.py (hunk 12 above)
Comment (tjaalton) [reply ipaplatform/fedora/paths.py:40]: yeah, you're right, this is fine

File: ipaserver/install/dnskeysyncinstance.py [hunk 13]
```diff
@@ -164,6 +165,87 @@ def __setup_dnssec_containers(self):
 
         self._ldap_mod("dnssec.ldif", {'SUFFIX': self.suffix, })
 
+    def __are_named_options_configured(self, options):
```
Comment (tiran) [inline ipaserver/install/dnskeysyncinstance.py:168]: Please no more double under prefix. They are annoying.

```suggestion
    def _are_named_options_configured(self, options):
```

File: ipaserver/install/dnskeysyncinstance.py [hunk 14]
```diff
@@ -164,6 +165,87 @@ def __setup_dnssec_containers(self):
...
+        """
+        Check whether sysconfig of named is patched or not.
+        Additional command line options for named are passed
```
Comment (tiran) [inline ipaserver/install/dnskeysyncinstance.py:171]:
```suggestion
        """Check whether sysconfig of named is patched

        Additional command line options for named are passed
```

File: ipaserver/install/dnskeysyncinstance.py [hunk 15]
```diff
@@ -164,6 +165,87 @@ def __setup_dnssec_containers(self):
...
+                'OPENSSL_ENGINE': constants.NAMED_OPENSSL_ENGINE,
+                'SOFTHSM_MODULE': paths.LIBSOFTHSM2_SO,
+            }
+            named_openssl_txt = ipautil.template_file(
+                os.path.join(paths.USR_SHARE_IPA_DIR,
+                             "bind.openssl.cnf.template"),
+                conf_file_dict)
+            with open(paths.DNSSEC_OPENSSL_CONF, 'w') as f:
+                os.fchmod(f.fileno(), 0o644)
+                f.write(named_openssl_txt)
```
Comment (tiran) [inline ipaserver/install/dnskeysyncinstance.py:204]: Which user and group needs read access to the file?

File: ipaserver/install/dnskeysyncinstance.py [hunk 16]
```diff
@@ -231,23 +305,17 @@ def __setup_softhsm(self):
...
-        named_fd.truncate(0)
-        named_fd.write(pin)
-        named_fd.close()
-        os.chmod(paths.DNSSEC_SOFTHSM_PIN, 0o770)
-        # chown to ods:named
-        os.chown(paths.DNSSEC_SOFTHSM_PIN, self.ods_uid, self.named_gid)
+        with open(paths.DNSSEC_SOFTHSM_PIN, 'w') as f:
+            # chown to ods:named
+            os.fchown(f.fileno(), self.ods_uid, self.named_gid)
+            os.fchmod(f.fileno(), 0o770)
```
Comment (tiran) [inline ipaserver/install/dnskeysyncinstance.py:311]: Why does this file have executable permission? The original code looks wrong. It's not an executable file.

```suggestion
            os.fchmod(f.fileno(), 0o660)
```

File: ipaserver/install/dnskeysyncinstance.py (hunk 13 above)
Comment (stanislavlevin) [reply ipaserver/install/dnskeysyncinstance.py:168]: I'm trying to follow the existing code style. Of course, I don't plan to annoy you.

Comment (stanislavlevin) [reply ipaserver/install/dnskeysyncinstance.py:168]: Fixed.

File: ipaserver/install/dnskeysyncinstance.py [hunk 17]
```diff
@@ -164,6 +165,87 @@ def __setup_dnssec_containers(self):
...
+        supplied by a vendor, at least, the base parsing of such
+        is required.
+        Current named command line options:
+        NS_MAIN_ARGS "46A:c:C:d:D:E:fFgi:lL:M:m:n:N:p:P:sS:t:T:U:u:vVx:X:"
+        If there are several same options the last passed wins.
+        """
+        if options:
+            pattern = r"[ ]*-[a-zA-Z46]*E[ ]*(.*?)(?: |$)"
+            p = re.compile(pattern)
+            engines = re.findall(p, options)
```
[the remark this answers is not in the document]
Comment (stanislavlevin) [reply ipaserver/install/dnskeysyncinstance.py:182]: I see,
Lib/re.py:
```python
231 def findall(pattern, string, flags=0):                                          
232     """Return a list of all non-overlapping matches in the string.              
233                                                                                 
234     If one or more capturing groups are present in the pattern, return          
235     a list of groups; this will be a list of tuples if the pattern              
236     has more than one group.                                                    
237                                                                                 
238     Empty matches are included in the result."""                                
239     return _compile(pattern, flags).findall(string)
```
Thank you.

[the remark this answers is not in the document]
Comment (stanislavlevin) [reply ipaserver/install/dnskeysyncinstance.py:182]: Fixed

**Comment by abbra** [comment]
We definitely want this to be completed.

**Comment by abbra** [comment]
@stanislavlevin could you please rebase it and address @tiran comments?

File: ipaserver/install/dnskeysyncinstance.py (hunk 15 above)
Comment (stanislavlevin) [reply ipaserver/install/dnskeysyncinstance.py:203]: current ownership and permissions:
```
# ls -la /etc/ipa/dnssec/
total 28
drwxr-xr-x 1 root root 4096 Aug 10 09:33 .
drwxr-xr-x 1 root root 4096 Aug 10 09:34 ..
-r--r----- 1 root ods   524 Aug 10 09:33 ipa-dnskeysyncd.keytab
-rw-r--r-- 1 root root  267 Aug 10 09:33 openssl.cnf
-rw-r--r-- 1 root root  145 Aug 10 09:33 softhsm2.conf
-r-------- 1 root root   30 Aug 10 09:33 softhsm_pin_so
```

Consumers of this file are:
- ipa-dnskeysyncd
- named/named-pkcs11

```
systemctl cat ipa-dnskeysyncd
# /usr/lib/systemd/system/ipa-dnskeysyncd.service
[Unit]
Description=IPA key daemon

[Service]
EnvironmentFile=/etc/sysconfig/ipa-dnskeysyncd
ExecStart=/usr/libexec/ipa/ipa-dnskeysyncd
User=ods
Group=named
SupplementaryGroups=ods
PrivateTmp=yes
...
```

```
# systemctl cat named
# /usr/lib/systemd/system/named.service
[Unit]
Description=Berkeley Internet Name Domain (DNS)
Wants=nss-lookup.target
Wants=named-setup-rndc.service
Before=nss-lookup.target
After=named-setup-rndc.service
After=network.target

[Service]
Type=forking
Environment=NAMEDCONF=/etc/named.conf
EnvironmentFile=-/etc/sysconfig/named
Environment=KRB5_KTNAME=/etc/named.keytab
PIDFile=/run/named/named.pid

ExecStartPre=/bin/bash -c 'if [ ! "$DISABLE_ZONE_CHECKING" == "yes" ]; then /usr/sbin/named-che
ckconf -z "$NAMEDCONF"; else echo "Checking of zone files is disabled"; fi'
ExecStart=/usr/sbin/named -u named -c ${NAMEDCONF} $OPTIONS
ExecReload=/bin/sh -c 'if /usr/sbin/rndc null > /dev/null 2>&1; then /usr/sbin/rndc reload; els
e /bin/kill -HUP $MAINPID; fi'

ExecStop=/bin/sh -c '/usr/sbin/rndc stop > /dev/null 2>&1 || /bin/kill -TERM $MAINPID'

PrivateTmp=true
...
```

It seems that below would enough:
```
# ls -la /etc/ipa/dnssec/openssl.cnf 
-rw-r----- 1 root named 267 Aug 10 09:33 /etc/ipa/dnssec/openssl.cnf
```

File: install/share/bind.openssl.cnf.template [hunk 18]
```diff
@@ -0,0 +1,14 @@
+# OpenSSL configuration file
+# File generated by IPA instalation
+openssl_conf = openssl_init
+
+[openssl_init]
+engines = engine_section
```
[the remark this answers is not in the document]
Comment (stanislavlevin) [reply install/share/bind.openssl.cnf.template:6]: @tiran, are you sure about RHEL? because crypto policy for named is only included for Fedora.

**Comment by stanislavlevin** [comment]
Rebased, I don't sure about the inclusion of SSL/TLS crypto policy for signing on private key.

File: install/share/bind.openssl.cnf.template (hunk 18 above)
[the remark this answers is not in the document]
Comment (abbra) [reply install/share/bind.openssl.cnf.template:6]: We need to follow (or obey) system-wide crypto policy. This means, for example, we are ensuring only ciphers allowed in the policy would be used.

[the remark this answers is not in the document]
Comment (stanislavlevin) [reply install/share/bind.openssl.cnf.template:6]: I have no objections about crypto policy, my question is : which one crypto policy to include for digital signature?
```
# cat /etc/crypto-policies/back-ends/opensslcnf.config
CipherString = @SECLEVEL=1:kEECDH:kRSA:kEDH:kPSK:kDHEPSK:kECDHEPSK:-aDSS:-3DES:!DES:!RC4:!RC2:!IDEA:-SEED:!eNULL:!aNULL:!MD5:-SHA384:-CAMELLIA:-ARIA:-AESCCM8
Ciphersuites = TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:TLS_AES_128_CCM_SHA256
MinProtocol = TLSv1
MaxProtocol = TLSv1.3
SignatureAlgorithms = ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:ed25519:ed448:rsa_pss_pss_sha256:rsa_pss_pss_sha384:rsa_pss_pss_sha512:rsa_pss_rsae_sha256:rsa_pss_rsae_sha384:rsa_pss_rsae_sha512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:RSA+SHA224:ECDSA+SHA1:RSA+SHA1
```
According to docs:
- https://www.openssl.org/docs/manmaster/man5/config.html
- https://www.openssl.org/docs/manmaster/man3/SSL_CONF_cmd.html
 these settings are related to SSL Configuration.

On the other hand, there is the bind-specific policy (included in named.conf):
```
# cat /etc/crypto-policies/back-ends/bind.config 
disable-algorithms "." {
RSAMD5;
DSA;
};
disable-ds-digests "." {
GOST;
};
```

**Comment by stanislavlevin** [comment]
Fixed upgrade.

**Comment by stanislavlevin** [comment]
```
2020-08-11 16:50:44,856 CRITICAL  Terminating execution, runtime exceeded 3600s
2020-08-11 16:50:44,859 CRITICAL  No free disk space
```
Re-run.

**Comment by stanislavlevin** [comment]
@tiran, please, let me know which crypto policy to use in this case? `/etc/crypto-policies/back-ends/opensslcnf.config` ?

**Comment by stanislavlevin** [comment]
Included crypto policy.

**Comment by abbra** [comment]
@stanislavlevin please remove the temp commit.

**Comment by stanislavlevin** [comment]
Removed the temp commit.

**Comment by abbra** [comment]
The code looks good to me. I am considering to merge this to master.

**Comment by stanislavlevin** [comment]
@abbra, thank you!

**Comment by abbra** [comment]
master:

* 173cd9b91b1048e852dca9c39dbbd38304367dd1 spec: Require ldns-utils
* a9334ce5e5fbedfb22a5779ce90e367c13979747 named: Remove no longer used paths
* 5c907e34ae4496482151cd9c8271d9931ac4056d named: Allow using of a custom OpenSSL engine for BIND
* bed09b7f85a2abf73e2dfec58ea8a094ae847d29 DNSKeySyncInstance: Populate named/ods uid/gid on instantiation
* 85ed106d78b0393aac7929cd2713f1097fb8a67d upgrade: Handle migration of BIND OpenSSL engine
* 721435cf7f2ed41fe807c34022fed31c792b4497 named: Make use of 'pkcs11' OpenSSL engine for BIND on Fedora31
* 53b341f94b2ba5f39696995e4ad772ffe757e495 spec: Bump required openssl-pkcs11 and softhsm
* 8716881fc41909d0ab5a32d2db79313d7b81ee49 service: Allow service to clean up its state
* ecfaf897b937e2333296d32960b2cc13843048f9 named: Don't override custom command line options for named
* e2030b8cad9eaf8760ec107f370be3b71ffbe2f4 named: Include crypto policy in openssl config
* 92157bc880600be137c634d5def788c1bf8c5e8a ipa-dnskeysyncd: Raise loglevel to DEBUG
